### PR TITLE
GenEra Conda Recipes Submission

### PR DESCRIPTION
**Pull request by AnupamGautam**
Dear Administrator,

Please find attached the Conda recipes for GenEra. Let us know if any further modifications are required. We look forward to your feedback.

Best regards,
Anupam

### CODE REVIEW

File: recipes/genera/meta.yaml
Comment (aliciaaevans) [inline recipes/genera/meta.yaml:16]: Bioconda only builds Python 3.8 and up.

File: recipes/genera/build.sh
Comment (aliciaaevans) [inline recipes/genera/build.sh:10]: Don't use git clone. This is a separate tool, so please add it as a separate recipe.

**Review by aliciaaevans — CHANGES_REQUESTED** [review]

File: recipes/genera/build.sh
Comment (aliciaaevans) [reply recipes/genera/build.sh:12]: You are right that the git clone command should be removed, but instead of adding the missing dependency as a source, a separate recipe should be created for the dependency in a new PR and then the dependencies updated in this recipe.

Comment (AnupamGautam) [reply recipes/genera/build.sh:12]: Hi @aliciaaevans, thank you for the review. This particular script doesn’t have a release version as I believe it’s part of a publication’s codebase. I’ve submitted a request on their GitHub for a release, but if they don’t create one, would it be possible to proceed with `git clone` despite the recommendation to avoid it?

File: recipes/genera/meta.yaml
Comment (AnupamGautam) [reply recipes/genera/meta.yaml:16]: Thank you will update it.

File: recipes/genera/build.sh
Comment (AnupamGautam) [reply recipes/genera/build.sh:10]: I’ve submitted a request on their GitHub for a release, but if they don’t create one, would it be possible to proceed with git clone despite the recommendation to avoid it?

Comment (martin-g) [reply recipes/genera/build.sh:10]: You can use 'git_url: ...' and 'git_rev: commit_sha' for it

Comment (AnupamGautam) [reply recipes/genera/build.sh]: Dear @aliciaaevans and @martin-g,

I have incorporated the suggested changes. Please let me know if any further adjustments are needed.

Best regards,  
Anupam

Comment (AnupamGautam) [reply recipes/genera/build.sh]: Dear @aliciaaevans and @martin-g,

The author of abSENSE has made a new release, so I’ve added it. However, I’m not certain if they would like a Conda package for it. I’ve reached out to confirm, and if they are interested, I can create a PR for the Conda package as well.

Best regards,
Anupam

**Comment by AnupamGautam** [comment]
@BiocondaBot please add label

**Comment by AnupamGautam** [comment]
Dear @aliciaaevans and @martin-g,

Could you please review the updated recipe and assist with the merge? Apologies for interrupting during a busy schedule, and thank you in advance for your help!

Best,
Anupam

File: recipes/genera/meta.yaml
Comment (bgruening) [inline recipes/genera/meta.yaml:40]: Please use SDPX identifiers

File: recipes/genera/meta.yaml
Comment (bgruening) [inline recipes/genera/meta.yaml:21]: please do not spin that strict, this is not how conda works. Strict pins can be used in environments.yml files, but packages should specify pins when they are really incompatible.

File: recipes/genera/meta.yaml
Comment (bgruening) [inline recipes/genera/meta.yaml:13]: This package should be noarch generic I think

File: recipes/genera/meta.yaml
Comment (AnupamGautam) [reply recipes/genera/meta.yaml]: added

File: recipes/genera/meta.yaml
Comment (AnupamGautam) [reply recipes/genera/meta.yaml:13]: Added.

**Comment by mfansler** [comment]
My recommendation is to use the newer `r-phytools` if possible. I almost never recommend for *recipes*<sup>**1**</sup> to use exact (`=`) specifications for dependencies. Only in cases with strong reasons should that be done.

Besides, Bioconda has adopted a `run_exports` strategy that serves to safeguard against using versions that deviate too far from those used to build the package. That doesn't, however, extend to Conda Forge packages, but one can effectively mimic it by providing bounds.

Otherwise, a proper solution to using either (1) older Bioconda packages from feedstocks that migrated to Conda Forge or (2) new Bioconda packages from feedstocks that migrated from Conda Forge has not been agreed upon. There is [a proposal](https://github.com/conda-forge/cfep/pull/52) to solve this from the Conda Forge side, but it still needs more feedback from Bioconda devs. Alternatively, Bioconda could allow recipes to use flexible solves.<sup>**2**</sup>

---

<sub>**[1]**: Environments on the other hand that go into pipelines should be exact, at least up to (semantically) minor versions.</sub>
<sub>**[2]**: My hot take on this is that `strict` channel priority is outdated and `flexible` is sufficiently reliable.</sub>

**Comment by AnupamGautam** [comment]
Hi @mfansler,

Thank you for your help. As suggested, I am working on loosening the version requirement. I will test it locally, and if it works, I will update the recipe accordingly.

Thank you again for your support.

Best regards,
Anupam

File: recipes/genera/meta.yaml
Comment (mfansler) [inline recipes/genera/meta.yaml:37]: It would be better to see the results in the logs. Also, there are 7 executables being moved into `CONDABIN` - they should also be tested.

```suggestion
    - genEra -h
```

File: recipes/genera/meta.yaml
Comment (mfansler) [inline recipes/genera/meta.yaml:10]: Is there a reason this is not a separate recipe?

File: recipes/genera/meta.yaml
Comment (mfansler) [inline recipes/genera/meta.yaml:6]: Version could be a variable. Here's canonical boilerplate for this:

```suggestion
{% set version = '1.4.2' %}

package:
   name: genera
   version: {{ version|replace('-', '_') }}

source:
  - url: https://github.com/josuebarrera/GenEra/archive/refs/tags/v{{ version }}.tar.gz
```

**Review by mfansler — CHANGES_REQUESTED** [review]

File: recipes/genera/meta.yaml
Comment (AnupamGautam) [reply recipes/genera/meta.yaml:12]: Dear @mfansler,  

I have implemented your suggestions and added a new test as well. Regarding the tool *abSENSE*, I have reached out to the author to confirm their interest in having a Conda package. If they express interest, I can create a PR for the package.  

For now, we can proceed like this, and if they author want I can create a PR later ,and update this package also. Could you please approve the workflow so that it can start? Thank you so much for your help!  

Best regards,  
Anupam

Comment (mfansler) [reply recipes/genera/meta.yaml:12]: As far as I see, abSENSE has no licensing declared, which is prohibitive to bundling it for Bioconda. Please ask the author to declare a license and include that in a release.

Once there is a license it should be moved to a separate package here. Note that the author does not have to be a maintainer here.

Comment (mfansler) [reply recipes/genera/meta.yaml:12]: Thanks for the additional information!

While the article itself is CC-BY-4.0, a copy of the code does not appear deposited with the article and thus it is not covered. If they want to extend exactly the same license, they could put a copy of [the CC-BY-4.0 text](https://creativecommons.org/licenses/by/4.0/legalcode.txt) in the repository somewhere.

Otherwise, unfortunately "open-source" doesn't translate to permission to redistribute.

Comment (AnupamGautam) [reply recipes/genera/meta.yaml]: Here’s a polished version of your message:  

---

**Subject:** Update on Package and Request for Workflow Initiation  

Dear @mfansler,  

The authors have added a license file to their repository and released a new version. I’ve updated the `meta.yaml` file to include this latest release. While they may still need time to decide whether they want an official Conda package, I can create one for them later if requested and make the necessary updates.  

For now, we can proceed with the current setup. Could you help initiate the workflow and assist in merging this?  

Thank you for your continued support!  

Best regards,  
Anupam

Comment (AnupamGautam) [reply recipes/genera/meta.yaml]: Dear @mfansler,

Thank you so much for your help—I really appreciate it. Let's wait and see what others have to say. If needed, I can also include this script in our package or create a dedicated Conda package for it.

Best regards,
Anupam

**Comment by AnupamGautam** [comment]
Dear @mfansler and @aliciaaevans,  

I have added the required recipe for **abSENSE** (it has been merged https://anaconda.org/bioconda/absense), removed the corresponding section from **GenEra**, and made the necessary adjustments. The updated recipe has passed all checks.  

I believe this issue can now be resolved and merged. Please let me know if further changes are needed.  

Best regards,  
Anupam

File: recipes/genera/meta.yaml
Comment (mfansler) [reply recipes/genera/meta.yaml:32]: I agree with the bot here. Are the exact specifications really necessary?

The `run_exports` of other Bioconda packages will add upper bounds that help constrain co-installation with too new versions in the future.

Also, [Python 3.8 has reached EOL](https://devguide.python.org/versions/#python-release-cycle), so Python 3.9 would make more sense as a minimum.

The `pip` and `setuptools` suggested by the bot can be ignored - they aren't relevant since this only copies scripts.

File: recipes/genera/build.sh
Comment (AnupamGautam) [reply recipes/genera/build.sh]: I have added the required recipe for abSENSE (it has been merged https://anaconda.org/bioconda/absense), removed the corresponding section from GenEra, and made the necessary adjustments

**Review by mfansler — CHANGES_REQUESTED** [review]

**Comment by AnupamGautam** [comment]
Dear @mfansler,  

I have adjusted the flexibility wherever possible. However, Python 3.8 (or versions below 3.9) is currently required as **abSENSE** is not compatible with Python >=3.9. Until the author provides an update, we will need to maintain this version requirement. Once they implement compatibility, I will update the recipe accordingly.  

Regarding other dependencies, some flags are only available in specific versions, which necessitates retaining version-specific requirements for now.  

Best regards,  
Anupam

**Review by mfansler — APPROVED** [review]
Okay, let's merge.

---

**Note**

> I have adjusted the flexibility wherever possible. However, Python 3.8 (or versions below 3.9) is currently required as **abSENSE** is not compatible with Python >=3.9. Until the author provides an update, we will need to maintain this version requirement. Once they implement compatibility, I will update the recipe accordingly.

Dependencies do not usually dictate the Python compatibility of a package. While that might restrict the environments in which it can be installed, the Python compatibility should be intrinsic to the package code (e.g., what syntactic features it uses). In this line, were an `abSENSE` that is Python v3.9+ released, one would immediately have the ability to co-install with it, rather than having to build and upload new artifacts.

However, I won't insist on this.

**Comment by AnupamGautam** [comment]
Dear @mfansler,

Thank you so much for your approval and support; I truly appreciate it. I believe I’ll need to wait for @aliciaaevans before the changes can be merged.

Best regards,  
Anupam

**Comment by aliciaaevans** [comment]
@BiocondaBot please fetch artifacts

**Review by aliciaaevans — APPROVED** [review]

**Pull request merged**
